woocommerce/woocommerce-android
# Background Image Upload: new logic for handling background work

Part of #4685 

### Description
First of all, sorry for the long PR, I tried to split the work on two PRs (#4701 and this one), but then I found that it just adds overhead without making the second PR any smaller.
The initial plan for supporting background image uploading (ref p91TBi-5SB-p2) was to keep the `ProductImagesService`'s logic, but after working on it (PR #4701), the solution was too fragile, and we can't ship it, mainly for two reasons:
1. We use `ACTION_GET_CONTENT` for retrieving local files, so we can't get a persistable read permission on those URIs, but since we depend on the queuing logic of `JobServiceIntent`, we won't `fetch` the files until the upload of previous ones finishes, which meant we will be getting error `SecurityException`s sometimes if the provider decided that we don't have permission to read it.
2. Depending on `JobServiceIntent` meant that sometimes the system may close the service and restarted later, which messes with our upload statuses.

So I changed the plan to have our in-house solution for queuing and handling work, and use a simple `Foreground Service` as a façade to let the system know that there is still work to do. I'll try to explain a bit about the solution:
- `MediaUploadHandler` acts as an orchestrator for all the upload process:
    - It receives requests for queuing uploads.
    - Keeps track of ongoing uploads and current errors.
    - Observes `ProductImagesUploadWorker` events: handles them, and forwards the relevant ones to the ViewModels.
    - Post notification about successes and failures.
- `ProductImagesUploadWorker` is responsible for executing work requests from `MediaUploadHandler`:
    - Queues the work requests and handles them.
    - Manages the `Foreground Service` status.
    - Updates the foreground notification content.
    - To update a product: we start by fetching it, and we retry in case of any errors for three times.

### Important
Background image uploading doesn't work currently for:
1. **Variations**, IMO it's not very important, since variations support only one image, and adding support for this one would require a lot of work to differentiate between product and variations.
2. **Adding product flow**: this is going to be handled in a separate PR later, since we need an additional logic to update the product ID after saving the product.

### Testing instructions
Start with the test cases from the PR https://github.com/woocommerce/woocommerce-android/pull/4684, to confirm that there is no regression for foreground uploading.

##### Single product:
###### Happy path
1. Open product details.
2. Pick some images from your device's memory.
3. Confirm that the image uploading starts.
4. Leave the product details screen (either by swiping up the app from the recent apps list, or using the back button)
5. Confirm that the upload continues.
6. Confirm that the product is updated afterward, and you can see a notification about a successful update.
7. Click on the notification.
8. Confirm that product screen is opened.

###### Unhappy path
To simulate a failure, use a proxy app like Charles Proxy, or turn on airplane mode after the upload begins, or use unsupported files like `HEIC`
1. Open product details.
2. Pick some images from your device's memory.
3. Confirm that the image uploading starts.
4. Leave the product details screen (either by swiping up the app from the recent apps list, or using the back button)
5. Once the upload fails, confirm that a notification is displayed.
6. Click on the notification.
7. Confirm that you are taken to the error details screen.

###### Return to product screen
1. Open product details.
2. Pick some images from your device's memory.
3. Confirm that the image uploading starts.
4. Leave the product details screen using back button.
5. Wait until uploading of some images finishes.
6. Go back to the product details.
7. Confirm that the uploaded images are displayed, and that `Save` button is visible.

###### Discard changes
1. Open product details.
2. Pick some images from your device's memory.
3. Confirm that the image uploading starts.
4. Go back to product details.
5. Wait until the uploading of some images finishes.
6. Click on the back button.
7. Confirm that the discard changes dialog is displayed.
8. Select Discard.
9. Confirm that the upload is canceled and that the service notification is removed.

##### Multiple products:
###### Happy path
1. Open product 1 details.
2. Pick some images from your device's memory.
3. Confirm that the image uploading starts.
4. Leave the product 1 details screen using back button.
5. Open product 2 details.
6. Pick some images from your device's memory.
7. Leave the product details screen.
8. Confirm that the uploading continues for both products.
9. Confirm that upload success notification is displayed for both products.

###### Unhappy path
Repeat the same steps as `Happy path` case, but using images that will fail, and confirm that failure notifications will be displayed.

### Images/gif
##### Single product (success):
https://user-images.githubusercontent.com/1657201/132214545-5614aa5c-1272-4583-8a0c-50fdfee7f2b5.mp4

##### Multiple products (success)
https://user-images.githubusercontent.com/1657201/132214583-06a1698c-f15e-4980-bf7b-515124be709e.mp4

##### Cancel upload
https://user-images.githubusercontent.com/1657201/132214644-e7e37a96-150e-418d-9682-e0b053191990.mp4

##### Upload failure
https://user-images.githubusercontent.com/1657201/132214688-50afc674-40c3-4ae4-ae22-f3254cf4d42b.mp4
